### ISSUE TITLE
Select single user server port on remote host

### ISSUE THREAD

**Issue by rkdarst** [jupyterhub/jupyterhub#1830]
This issue was submitted to batchspawner some time ago and they proposed it might be useful here:

https://github.com/jupyterhub/batchspawner/pull/58

Abstract from there:

> During last week lab period, two users reported being unable to connect to their notebook. After inspecting their notebook log, I found this message:
>
>>    [I 2018-02-16 10:26:57.826 SingleUserNotebookApp notebookapp:1191] The port 58824 is already in use, trying another port.
>>    [C 2018-02-16 10:26:57.826 SingleUserNotebookApp notebookapp:1203] ERROR: the notebook server could not be started because no available port could be found.
>
> The users could not access their notebook, because the singleserver could not start. The singleserver could not start because it was assigned a port number that was already used on the compute, probable by another singleserver used by another student.

and solution delays choosing the port until the singleuser server starts.

I haven't tested and I'm not involved there... but I could try applying here.

**Comment by minrk** [comment]
Ah, yes. singleuser notebook shouldn't allow retrying ports. That should be fixable here.

**Comment by mbmilligan** [comment]
Oh yes, I *was* supposed to have pushed this up the chain for discussion. The reason we didn't adopt the PR there was that we wanted to propose the new API as a generally-useful feature for the Hub to support, probably gated by a configuration setting. Batchspawner is certainly not the only spawner where it could be useful to optionally let the singleuser server choose the port number.

**Comment by manics** [comment]
Is this still an issue?

**Comment by mbmilligan** [comment]
> Is this still an issue?

Yes. Batchspawner did eventually merge the PR linked in the first post, and Kubespawner is considering adding a similar API callback. PR #2727 exists as a proposal to bring this functionality into the core codebase.
